### ISSUE TITLE
Allow custom headers on no-cors requests

### ISSUE THREAD

**Issue by RReverser** [whatwg/fetch#380]
*(posting this as a result of Twitter discussion with @annevk)*

Current state:
 
Most of the page resource requests are `no-cors` by default, even if to the same origin.

If `no-cors` request was made to the same origin, you can read its response, however you can't set custom request headers. In particular, it makes hard to add specific metadata in Service Worker to be used by server-side for customized response that only Service Worker can handle.

This is possible to workaround by passing custom values under the allowed header names, but it would be unclean and pretty limited solution.

Another option is to (re)issue request with `cors` or `same-origin` mode in addition to modified headers if origin matches, but then you still opt into response errors should the origin decide to redirect to 3rd-party, which in unfavorable when writing Service Worker code that can be inserted to any website. So you still need to handle failure case and reissue original request should CORS error happen, which results in double-trip and weakens performance. (This all applies to adding `crossorigin` attribute to page resources as well).

If I understand correctly, the motivation for silently ignoring addition of custom headers to `no-cors` requests even to the same origin is exactly the fact that origin can redirect to 3rd party and browser would have to make that redirect with same custom headers, which now would violate CORS rules and is unacceptable from security point.

My suggestion is: allow addition of custom headers to such requests instead of silently ignoring them, but if the origin attempts to redirect to a 3rd-party, remove the disallowed headers from the further request hops (issued to the new origin).

This would simplify adding custom headers on page resources by 1) not ignoring headers at time of addition and 2) not forcing opt-in into errors caused by CORS.

This also doesn't weaken security - as was shown above, workarounds with double-trip with different modes are already possible, but they are less performant.

Thoughts?

**Comment by annevk** [comment]
A new potential attack I see is that if an attacker somehow gets to add request headers and those request headers do something at the redirect level (or turn a response into a redirect) there's a new vulnerability of sorts that you cannot have today. Not sure how much of a deal-breaker that should be.

**Comment by RReverser** [comment]
@annevk Can you please provide any examples where an attacker could achieve that with request headers? Also, wouldn't it work in case of `cors` or `same-origin` modes as well?

**Comment by annevk** [comment]
The server would have to do something with request headers, like adding debug information to a redirect URL. "same-origin" would block at a cross-origin URL. "cors" would require a preflight due to the request headers.

**Comment by RReverser** [comment]
Ah I see. Although it sounds like in this case server explicitly wants to change redirect URL based on metadata it received in own request headers, so doesn't seem like vulnerability, rather as a part of the feature (custom response based on request headers).

**Comment by annevk** [comment]
I'm inclined to close this as "no-cors" is already a source of issues, we want to clamp it down further because of Spectre, and it seems unlikely implementers would be interested in adding complexity because of those two factors. Hope that's understandable.

**Comment by annevk** [comment]
Okay, let's get a second opinion from @jakearchibald before we move onward. (If you know someone else who could chime in that would be good too. Bit hard to estimate the size of the problem.)

**Comment by wanderview** [comment]
So, when the fetch spec was originally written it had a lot of "if a request is created with a potentially bad condition throw or discard some bad bits immediately" type logic.  Lately we have moved towards waiting to trigger the error handling until the request actually ends up trying to trigger a cross-origin request.

For example, #663 does not forbid "manual" redirect mode for "no-cors" requests.  It just makes any such fetch that goes cross-origin reject.

The suggestion in https://github.com/whatwg/fetch/issues/380#issue-174581060 seems consistent with this approach.  The request can be created and function with full features for same-origin URLs.  When it goes cross-origin the data is stripped (or we could reject).

**Comment by jakearchibald** [comment]
If we want this, I agree with @wanderview. I guess it'd be a new mode like `no-cors-same-origin`, which would allow more headers, but if request URL is/becomes cross-origin, downgrade to `cors` and filter the headers accordingly.

**Comment by RReverser** [comment]
If it's a new mode, it somewhat defeats the purpose to be able to intercept/modify regular resource requests from the page, and would only apply to explicit `fetch(...)` with that mode specified. Or are you suggesting that such new mode would be the new default for any subresources?

**Comment by RReverser** [comment]
Oh, or do you mean that one would be able to rewrite request and set a new `mode` in a SW? If that won't throw for third-party requests either (as far as I remember, it shouldn't anymore?), then yeah, should work.

**Comment by jakearchibald** [comment]
> Oh, or do you mean that one would be able to rewrite request and set a new mode in a SW?

Yeah, that's what I was thinking. Changing the default mode or the behaviour of `no-cors` feels risky, and I'm not sure how we'd evaluate that risk.

**Comment by RReverser** [comment]
Sounds good to me.

As long as there is a way to make such "optimistic" modifications to the request that will be guaranteed to never fail, but rather just be dropped on third-party redirects etc., everything else seems like an implementation detail.

**Comment by wanderview** [comment]
> Changing the default mode or the behaviour of no-cors feels risky, and I'm not sure how we'd evaluate that risk.

Adding another mode complicates the platform and API surface, though.

Today we have:

1. Create no-cors request
2. Add custom header to request
3. Custom header is thrown away
4. Request fetches cross-origin
5. Server does not see custom header

If we moved where the stripping occurs, we would have:

1. Create no-cors request
2. Add custom header to request
3. Request fetches cross-origin
4. Custom header is thrown away
5. Server does not see custom header

The end result is that the server does not see it.  This seems much nicer than adding yet another mode and is consistent with how we are building other no-cors restrictions in fetch.

I'd be curious what risks people have in mind that would prevent this.

**Comment by wanderview** [comment]
To clarify, I'm suggesting we modify step 5 of main fetch here:

https://fetch.spec.whatwg.org/#main-fetch

To say something like:

```
request’s mode is "no-cors"
  1. If request’s redirect mode is not "follow", then return a network error.
  2. Set request’s response tainting to "opaque".
  3. Remove all headers from request that are not a CORS-safelisted request-header.
  4. Return the result of performing a scheme fetch using request.
```

**Comment by annevk** [comment]
Presumably that coupled with removing the checks in the `Headers` object, which I further tightened earlier today. I'd be okay with that I think, provided someone writes exhaustive tests that ensure it gets implemented that way.
